doc/source/guide/in_depth/code/functional_indepth_example.py

```diff
@@ -0,0 +1,124 @@
+import odl
+
+
+# Here we define the functional
+class MyFunctional(odl.solvers.Functional):
+
+    """This is my functional: ``||x||_2^2 + <x, y>``."""
+
+    def __init__(self, space, linear_term):
+        """Initialize a new instance."""
+        # This comand calls the init of Functional and sets a number of
+        # parameters associated with a functional. All but domain have default
+        # values if not set.
+        super().__init__(space=space, linear=False, grad_lipschitz=2)
+
+        # We need to check that linear_term is in the domain. Then we store the
+        # value of linear_term for future use.
+        if linear_term not in space:
+            raise TypeError('linear_term is not in the domain!')
+        self.linear_term = linear_term
+
+    # Defining the _call function. This method is used for evaluation of
+    # the functional and always needs to be implemented.
+    def _call(self, x):
+        """Evaluate the functional."""
+        return x.norm()**2 + x.inner(self.linear_term)
+
+    # Next we define the gradient. Note that this is a property.
+    @property
+    def gradient(self):
+        """The gradient operator."""
+
+        # First we store the functional in a variable
+        functional = self
+
+        # The class corresponding to the gradient operator.
+        class MyGradientOperator(odl.Operator):
+
+            """Class implementing the gradient operator."""
+
+            def __init__(self):
+                """Initialize a new instance."""
+                super().__init__(domain=functional.domain,
+                                 range=functional.domain)
+
+                # Here we can access the stored functional and store a
+                # copy of it.
+                self.__functional = functional
+
+            def _call(self, x):
+                """Evaluate the gradient."""
+                return 2.0 * x + self.__functional.linear_term
+
+        return MyGradientOperator()
+
+    # Next we define the convex conjugate functional.
+    @property
+    def convex_conj(self):
+        """The convex conjugate functional."""
+        # This functional is implemented below.
+        return MyFunctionalConjugate(space=self.space, y=self.linear_term)
+
+
+# Here is the conjugate functional. Note that this is a separate class, in
+# contrast to the gradient which was implemented as an inner class. One
+# advantage with the inner class it that we don't have to pass as many
+# parameters when initializing, on the other hand having separate classes
+# normally improves readibility of the code. Both methods are use throughout
+# the odl package.
+class MyFunctionalConjugate(odl.solvers.Functional):
+
+    """Conjugate functional to ``||x||_2^2 + <x,y>``.
+
+    This funtional has the analytic expression
+
+        ``f^*(x) = ||x||^2/2 - ||x-y||^2/4 + ||y||^2/2 - <x,y>``.
+    """
+
+    def __init__(self, space, y):
```

Comment:
The `linear_term` guy has changed name to `y` here. I'd prefer if they stayed the same, and for an example like this `y` is perfectly ok.